#' @title RNA-seq example
#' @description Dummy example of a DESeq2 result for differential expression analysis on RNA-seq data
#' @format A data frame with 300 rows and 7 variables:
#' \describe{
#'   \item{\code{geneName}}{genes symbols}
#'   \item{\code{baseMean}}{The average of the normalized count values, dividing by size factors, taken over all samples}
#'   \item{\code{log2FC}}{the log2 value of the Fold Change expression between two conditions}
#'   \item{\code{lfcSE}}{log2 Fold Change standard error (SE)}
#'   \item{\code{stat}}{Wald statistic}
#'   \item{\code{pvalue}}{Wald test p-value}
#'   \item{\code{padj}}{BH adjusted p-values}
#'}
#' @source Simulated data
"RNAseq"
#'
#'
#'
#'
#' @title RNA-seq example
#' @description List result of the function \code{read.computeMatrix.file} used to read a matrix.gz file generated by deepTools computeMatrix function.
#' @format A named list of 3 variables:
#' \describe{
#'   \item{\code{metadata}}{data.frame with the information gotten from the matrix_file.gz}
#'   \item{\code{matrix.data}}{data.frame with the scores gotten from}
#'   \item{\code{original.file.path}}{with full path to the original matrix_file.gz}
#'}
#' @source \url{http://path.to.paper/}
"deeptools.matrix"
#'
#'
#'
#'
#' @title CNV data results example
#' @description Simulation of Copy Number Variation (CNV) analysis on a cohort of patients.
#' @format A data frame with 25 rows and 9 variables:
#' \describe{
#'   \item{\code{geneName}}{hypothetical gene symbols}
#'   \item{\code{patient_1 ... patient_N}}{hypothetical patients ID}
#'}
#' @source Simulated data
"CNV.data"
